stdlib/public/runtime/Casting.cpp

```diff
@@ -2755,6 +2805,165 @@ swift::swift_conformsToProtocol(const Metadata *type,
+        auto demangling = _swift_buildDemanglingForMetadata(metadata);
+
+        auto N = mangleNode(demangling);
+        if (N == typeName) {
+          foundMetadata = metadata;
+          break;
+        }
+      }
+    }
+  }
+
+  return foundMetadata;
+}
+
+// returns the generic type metadata for type named by boundGenericNode
+static const Metadata *
+_metadataForBoundGenericNode(const ConformanceState &C,
+                             const NodePointer boundGenericNode) {
+  if (boundGenericNode->getNumChildren() != 2)
+    return nullptr;
+
+  auto templateTypeNode = boundGenericNode->getChild(0);
+  auto typeListNode = boundGenericNode->getChild(1);
+
+  if (templateTypeNode->getKind() != Node::Kind::Type ||
+      typeListNode->getKind() != Node::Kind::TypeList)
+    return nullptr;
+
+  size_t numGenericArguments = typeListNode->getNumChildren();
+  void **arguments = new void *[numGenericArguments];
+
+  for (size_t i = 0; i < numGenericArguments; i++) {
+    auto metadata = _metadataForNode(C, typeListNode->getChild(i));
+    arguments[i] = const_cast<void *>(reinterpret_cast<const void *>(metadata));
+    if (arguments[i] == nullptr) {
+      delete arguments;
+      return nullptr;
+    }
+  }
+
+  unsigned sectionIdx = 0;
+  unsigned endSectionIdx = C.SectionsToScan.size();
+  auto templateName = mangleNode(templateTypeNode);
+  const Metadata *foundMetadata = nullptr;
+
+  for (; sectionIdx < endSectionIdx; ++sectionIdx) {
+    auto &section = C.SectionsToScan[sectionIdx];
+    for (const auto &record : section) {
+      if (record.getTypeKind()
+                   == ProtocolConformanceTypeKind::UniqueGenericPattern
+                 && record.getConformanceKind()
+                   == ProtocolConformanceReferenceKind::WitnessTable) {
+        auto R = record.getGenericPattern();
+
+        if (_genericPatternMatchesTemplate(R, llvm::StringRef(templateName))) {
+          foundMetadata =
+            swift::swift_getGenericMetadata(const_cast<GenericMetadata *>(R), arguments);
+          break;
+        }
+      }
+    }
+  }
+
+  delete arguments;
+
+  return foundMetadata;
+}
+
+static const Metadata *
+_metadataForNode(const ConformanceState &C, const NodePointer typeNode) {
+  if (typeNode->getKind() != Node::Kind::Type || !typeNode->hasChildren())
+    return nullptr;
+
+  auto childNode = typeNode->getChild(0);
+  const Metadata *foundMetadata = nullptr;
+
+  switch (childNode->getKind()) {
+  case Node::Kind::BoundGenericClass:
+  case Node::Kind::BoundGenericEnum:
+  case Node::Kind::BoundGenericStructure:
+    foundMetadata = _metadataForBoundGenericNode(C, childNode);
+    break;
+  default:
+    foundMetadata = _metadataForTypeNode(C, childNode);
+    break;
+  }
+
+  return foundMetadata;
+}
+
+/// Return the type metadata for a given mangled (_Tt) name.
+static const Metadata *
+_metadataForMangledName(const llvm::StringRef mangledName) {
+  if (!mangledName.startswith("_Tt"))
```

Comment:
I preferred to keep it there for similarity to the libobjc equivalent APIs but, not a biggie.